### PR TITLE
Xetla support 2024.2

### PR DESCRIPTION
**Pull request by sunjiweiswift**
## Type of Change

feature or bug fix or documentation or others  --   Update compiler API to 2024.2
API changed or not  --  Not

## Description
1 Update prefetch 1D API
2 Update atomic API
3 Update gather/scatter API
4 Update 1 fp16 load/store
5 Unify all length units to bytes instead of elems(common.hpp)

2D load/store/prefetch not updated, will be updated in subsequent PR

detail description 
Issues: xxx

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

**Comment by airMeng** [comment]
IPEX will only upgrade oneapi2024.2 after next week, let's hold on this

File: tests/unit/raw_send/kernel_func.hpp [hunk 1]
```diff
@@ -34,7 +34,8 @@ struct raw_send_with_2_source_and_1_destination_func {
     offsets += offset;
     xetla_vector<uint64_t, SIMD> dsec = offsets + (uint64_t)a;
 
-    xetla_vector<dtype, SIMD> A_load_vec = xetla_load_global(a, offsets);
+    xetla_vector<dtype, SIMD> A_load_vec =
+        xetla_load_global<dtype, SIMD, 1>(a, offsets);
```
Comment (JianpingChen066) [inline tests/unit/raw_send/kernel_func.hpp:38]: why now template argument can't be inferred from function arguments ?

File: include/common/core/memory.hpp [hunk 2]
```diff
@@ -233,78 +230,132 @@ constexpr __ESIMD_NS::atomic_op get_atomic_op(gpu::xetla::atomic_op ao) {
...
+    cache_hint L2H = cache_hint::cached>
 __XETLA_API void xetla_prefetch_global(
-    Ty* p,
-    xetla_vector<uint32_t, N> offsets,
-    xetla_mask<N> pred = 1) {
-  using T = native_type_t<Ty>;
+    T* p,
+    xetla_vector<uint32_t, N / VS> byte_offsets,
+    xetla_mask<N / VS> mask = 1) {
+#if __INTEL_LLVM_COMPILER >= 20240200
```
Comment (JianpingChen066) [inline include/common/core/memory.hpp:306]: any performance tests for the xetla instruction level modification ?

File: tests/unit/raw_send/kernel_func.hpp (hunk 1 above)
Comment (sunjiweiswift) [reply tests/unit/raw_send/kernel_func.hpp:38]: The third template parameter VS = 1 or non-1 is 2 different specializations. VS is a new template parameter added this time

File: include/common/core/memory.hpp (hunk 2 above)
Comment (sunjiweiswift) [reply include/common/core/memory.hpp:306]: For 1x4096x4096 and 32x4096x4096 1204x4096x4096 UT performance is the same.

Comment (sunjiweiswift) [reply include/common/core/memory.hpp:306]: The performance of a single instruction is not easy to test

File: tests/unit/raw_send/kernel_func.hpp (hunk 1 above)
Comment (sunjiweiswift) [reply tests/unit/raw_send/kernel_func.hpp:38]: Is there a third template parameter corresponding to 2 different template functions

**Review by DDEle — APPROVED** [review]
Can be merged as internal IPEX PR merged.

Still, some diff between IPEX to be close (maybe in another PR).

```diff
diff --git a/include/experimental/group/gemm/impl/int4_dequantize_xe.hpp b/include/experimental/group/gemm/impl/int4_dequantize_xe.hpp
index 06db3218..b91bcd74 100644
--- a/include/experimental/group/gemm/impl/int4_dequantize_xe.hpp
+++ b/include/experimental/group/gemm/impl/int4_dequantize_xe.hpp
@@ -512,9 +512,14 @@ class gemm_t<
 
     int scale_prefetch_addr_i = args.inner_loop_start;
     int tile_k_idx = args.inner_loop_start;
+    uint32_t prefetch_stages =
+        stages < args.inner_loop_count ? stages : args.inner_loop_count;
+    uint32_t prefetch_compute_stages =
+        stages < args.inner_loop_count ? args.inner_loop_count - stages : 0;
+    uint32_t compute_stages =
+        stages < args.inner_loop_count ? stages : args.inner_loop_count;
     SW_BARRIER();
-#pragma unroll
-    for (uint32_t i = 0; i < stages; i++) {
+    for (uint32_t i = 0; i < prefetch_stages; i++) {
       subgroup::tile_prefetch<cache_hint::cached, cache_hint::cached>(
           matA_prefetch_payload);
       subgroup::tile_prefetch<cache_hint::cached, cache_hint::cached>(
@@ -542,8 +547,7 @@ class gemm_t<
         }
       }
     }
-
-    for (uint32_t i = 0; i < args.inner_loop_count; i++) {
+    for (uint32_t i = 0; i < prefetch_compute_stages; i++) {
       if constexpr (enable_periodic_sync) {
         if ((i % sync_freq) == 0) {
           if constexpr (wg_size_x > 1) {
@@ -651,6 +655,78 @@ class gemm_t<
       }
     }
     SW_BARRIER();
+    for (uint32_t i = 0; i < compute_stages; i++) {
+      if constexpr (enable_periodic_sync) {
+        if ((i % sync_freq) == 0) {
+          if constexpr (wg_size_x > 1) {
+            nbarrier_a.arrive();
+          }
+          if constexpr (arch_tag >= gpu_arch::XeHpc) {
+            if constexpr (wg_size_y > 1) {
+              nbarrier_b.arrive();
+            }
+          }
+        }
+      }
+      subgroup::tile_load<cache_hint::cached, cache_hint::cached>(
+          matA, matA_payload);
+      subgroup::tile_load<cache_hint::cached, cache_hint::cached>(
+          matB, matB_payload);
+      // subgroup::tile_load<cache_hint::uncached, cache_hint::uncached>(
+      //     matB, matB_payload);
+      subgroup::tile_load<cache_hint::cached, cache_hint::cached>(
+          scale, scale_payload);
+      if constexpr (compute_policy::quant_mode != quant_mode::I4_SYM) {
+        subgroup::tile_load<cache_hint::cached, cache_hint::cached>(
+            zero_pt, zero_pt_payload);
+      }
+      tile_k_idx++;
+      SW_BARRIER();
+      matA_payload.template update_tdesc<update_dir_a>(matA_t::tile_size_x);
+      matB_payload.template update_tdesc<update_dir_b>(matB_t::tile_size_y);
+      if (tile_k_idx % scale_addr_update_freq == 0) {
+        scale_payload.template update_tdesc<update_dir_b>(scale_t::tile_size_y);
+      }
+      if constexpr (compute_policy::quant_mode != quant_mode::I4_SYM) {
+        if (tile_k_idx % zero_pt_addr_update_freq == 0) {
+          zero_pt_payload.template update_tdesc<tdesc_update_dir::y_dir>(
+              zero_pt_t::tile_size_y);
+        }
+      }
+      SW_BARRIER();
+      matA_acc_t matA_acc;
+      matB_acc_t matB_acc;
+      if constexpr (is_vnni_tiled_a) {
+        subgroup::vnni_reverse(matA);
+      }
+      subgroup::elemwise_cvt(matA_acc, matA);
+
+      dequantize(matB_acc, matB, scale, zero_pt, dequantize_args);
+      SW_BARRIER();
+      if constexpr (is_gemv) {
+        tile_mma::mma(
+            matAcc, matAcc, matC, matB_acc, matA_acc, i == compute_stages - 1);
+      } else {
+        if constexpr (is_col_major_b) {
+          tile_transpose(matB_acc);
+        }
+        tile_mma::mma(matC, matC, matB_acc, matA_acc);
+      }
+      SW_BARRIER();
+      if constexpr (enable_periodic_sync) {
+        if ((i % sync_freq) == 0) {
+          if constexpr (wg_size_x > 1) {
+            nbarrier_a.wait();
+          }
+          if constexpr (arch_tag >= gpu_arch::XeHpc) {
+            if constexpr (wg_size_y > 1) {
+              nbarrier_b.wait();
+            }
+          }
+        }
+      }
+    }
+    SW_BARRIER();
   }
 
  private:
diff --git a/include/subgroup/tile/impl/load_xe.hpp b/include/subgroup/tile/impl/load_xe.hpp
index ace2ed1a..1ce5e64d 100644
--- a/include/subgroup/tile/impl/load_xe.hpp
+++ b/include/subgroup/tile/impl/load_xe.hpp
@@ -500,7 +500,10 @@ tile_load(tile_t& tile, payload_t& payload) {
             load_elems,
             payload_t::vector_size,
             L1,
-            L2>(payload.base_ptr, channel_offset + address_offset, mask);
+            L2>(
+            payload.base_ptr,
+            payload.channel_offset + payload.base_offset + address_offset,
+            mask);
 
         if constexpr (
             payload_t::vector_size > 1 && payload_t::num_channel > 1) {
diff --git a/include/subgroup/tile/impl/payload_xe.hpp b/include/subgroup/tile/impl/payload_xe.hpp
index b39b132a..2316ef23 100644
--- a/include/subgroup/tile/impl/payload_xe.hpp
+++ b/include/subgroup/tile/impl/payload_xe.hpp
@@ -1655,12 +1655,11 @@ struct prefetch_payload_t<
         reg_layout_>,
     num_coop_sg_,
     arch_tag_,
-    std::enable_if_t<
-        (!arch_has_2d_load_store<arch_tag_>) &&
-        (((block_size_y_ != 1 || tile_size_y_ != 1) &&
-          mem_layout_ == mem_layout::row_major) ||
-         ((block_size_x_ != 1 || tile_size_x_ != 1) &&
-          mem_layout_ == mem_layout::col_major))>> {
+    std::enable_if_t<(!arch_has_2d_load_store<arch_tag_>)&&(
+        ((block_size_y_ != 1 || tile_size_y_ != 1) &&
+         mem_layout_ == mem_layout::row_major) ||
+        ((block_size_x_ != 1 || tile_size_x_ != 1) &&
+         mem_layout_ == mem_layout::col_major))>> {
   using dtype = native_type_t<dtype_>;
   using mem_desc_t =
       mem_desc_t<dtype_, mem_layout_, mem_space::global, alignment_>;
@@ -1902,10 +1901,9 @@ struct prefetch_payload_t<
         reg_layout_>,
     num_coop_sg_,
     arch_tag_,
-    std::enable_if_t<
-        (arch_has_2d_load_store<arch_tag_>) &&
-        (((tile_size_y_ != 1) && mem_layout_ == mem_layout::row_major) ||
-         ((tile_size_x_ != 1) && mem_layout_ == mem_layout::col_major))>> {
+    std::enable_if_t<(arch_has_2d_load_store<arch_tag_>)&&(
+        ((tile_size_y_ != 1) && mem_layout_ == mem_layout::row_major) ||
+        ((tile_size_x_ != 1) && mem_layout_ == mem_layout::col_major))>> {
   using dtype = dtype_;
   using mem_desc_t =
       mem_desc_t<dtype_, mem_layout_, mem_space::global, alignment_>;
```

**Comment by sunjiweiswift** [comment]
https://github.com/intel/neural-speed/pull/320

**Review by JianpingChen066 — APPROVED** [review]

**Pull request merged**